File: predict.c
#include "cnn_model.h"
#include "test_image_data.h"
#include <stdio.h>
#include <math.h> 

#define KERNEL_SIZE 3
#define NUM_KERNELS 8
#define DENSE_OUT 10
#define CONV_SIZE 26  // (IMG_SIZE - KERNEL_SIZE + 1)
#define IMG_SIZE 28
#define MAX_SIZE 128
#define DENSE_IN 5408  // (NUM_KERNELS * CONV_SIZE * CONV_SIZE)

void relu(float feature[CONV_SIZE][CONV_SIZE]);
void softmax(float x[DENSE_OUT]);
void convolve2d(const float image[IMG_SIZE][IMG_SIZE], const float kernel[KERNEL_SIZE][KERNEL_SIZE], float output[CONV_SIZE][CONV_SIZE]);
void flatten(float feature_maps[NUM_KERNELS][CONV_SIZE][CONV_SIZE], float flat[DENSE_IN]);
void dense_layer(const float W_dense[DENSE_OUT][DENSE_IN], const float flat[DENSE_IN], const float b_dense[DENSE_OUT], float result[DENSE_OUT]);
int predict(
    const float image[IMG_SIZE][IMG_SIZE],
    const float kernels[NUM_KERNELS][KERNEL_SIZE][KERNEL_SIZE],
    const float W_dense[DENSE_OUT][DENSE_IN],
    const float b_dense[DENSE_OUT],
    float probs[DENSE_OUT]   // output softmax probabilities
);


// Activation function on the 26 x 26 feature map produced from convolution layer
void relu(float feature[CONV_SIZE][CONV_SIZE]) {
    for (int i = 0; i < CONV_SIZE; i++) {
        for (int j = 0; j < CONV_SIZE; j++) {
            if (feature[i][j] < 0) {
                feature[i][j] = 0;
            }
        }
    }
}

void softmax(float x[DENSE_OUT]) {
    float max_val = x[0];
    for (int i = 1; i < DENSE_OUT; i++) {
        if (x[i] > max_val) {
            max_val = x[i];
        }
    }

    float sum = 0.0f;
    float exp_vals[DENSE_OUT];

    for (int i = 0; i < DENSE_OUT; i++) {
        exp_vals[i] = expf(x[i] - max_val);
        sum += exp_vals[i];
    }

    for (int i = 0; i < DENSE_OUT; i++) {
        x[i] = exp_vals[i] / sum;
    }
}


// === 2D valid convolution ===
// image: [28][28]
// kernel: [3][3]
// output: [26][26]
void convolve2d(const float image[IMG_SIZE][IMG_SIZE], const float kernel[KERNEL_SIZE][KERNEL_SIZE], float output[CONV_SIZE][CONV_SIZE]) {
    for (int i = 0; i < CONV_SIZE; i++) {
        for (int j = 0; j < CONV_SIZE; j++) {
            float sum = 0.0f;
            for (int ki = 0; ki < KERNEL_SIZE; ki++) {
                for (int kj = 0; kj < KERNEL_SIZE; kj++) {
                    sum += image[i + ki][j + kj] * kernel[ki][kj];
                }
            }
            output[i][j] = sum;
        }
    }
}

// === Flatten feature maps ===
// input: feature_maps[num_kernels][26][26]
// output: flat[DENSE_IN]
void flatten(float feature_maps[NUM_KERNELS][CONV_SIZE][CONV_SIZE], float flat[DENSE_IN]) {
    int index = 0;
    for (int k = 0; k < NUM_KERNELS; k++) {
        for (int i = 0; i < CONV_SIZE; i++) {
            for (int j = 0; j < CONV_SIZE; j++) {
                flat[index++] = feature_maps[k][i][j];
            }
        }
    }
}

// === Dense layer: Z = W * flat + b ===
// W_dense: [DENSE_OUT][DENSE_IN]
// flat: [DENSE_IN]
// b_dense: [DENSE_OUT]
// output: Z[DENSE_OUT]
void dense_layer(const float W_dense[DENSE_OUT][DENSE_IN], const float flat[DENSE_IN], const float b_dense[DENSE_OUT], float result[DENSE_OUT]) {
    for (int i = 0; i < DENSE_OUT; i++) {
        float sum = b_dense[i];
        for (int j = 0; j < DENSE_IN; j++) {
            sum += W_dense[i][j] * flat[j];
        }
        result[i] = sum;
    }
}

// === Predict function ===
// Returns index of max softmax prob, and fills probs output array
int predict(
    const float image[IMG_SIZE][IMG_SIZE],
    const float kernels[NUM_KERNELS][KERNEL_SIZE][KERNEL_SIZE],
    const float W_dense[DENSE_OUT][DENSE_IN],
    const float b_dense[DENSE_OUT],
    float probs[DENSE_OUT]   // output softmax probabilities
) {
    // 1) Convolve + relu for each kernel
    static float feature_maps[NUM_KERNELS][CONV_SIZE][CONV_SIZE];
    for (int k = 0; k < NUM_KERNELS; k++) {
        convolve2d(image, kernels[k], feature_maps[k]);
        relu(feature_maps[k]);
    }

    // 2) Flatten feature maps
    static float flat[DENSE_IN];
    flatten(feature_maps, flat);

    // 3) Dense layer
    static float Z[DENSE_OUT];
    dense_layer(W_dense, flat, b_dense, Z);

    // 4) Softmax
    for (int i = 0; i < DENSE_OUT; i++) probs[i] = Z[i];
    softmax(probs);
    // 5) Find max prob class
    int max_idx = 0;
    float max_val = probs[0];
    for (int i = 1; i < DENSE_OUT; i++) {
        if (probs[i] > max_val) {
            max_val = probs[i];
            max_idx = i;
        }
    }

    return max_idx;
}

int main() {
    float probs[DENSE_OUT];
    int predicted = predict(test_image, kernels, W_dense, b_dense, probs);

    printf("Predicted label: %d\n", predicted);
    printf("Actual label:    %d\n", test_label);

    printf("Softmax probabilities:\n");
    for (int i = 0; i < DENSE_OUT; i++) {
        printf("Class %d: %.5f\n", i, probs[i]);
    }
    return 0;
}

File: predict.py
import numpy as np
from scipy.signal import convolve2d
import re
# === Load CNN model from C header ===
def load_cnn_model(filename="cnn_model.h"):
    with open(filename, "r") as f:
        content = f.read()

    # === Parse kernels ===
    kernel_start = content.find("static const float kernels")
    if kernel_start == -1:
        raise ValueError("kernels block not found")

    kernel_block = content[kernel_start:]
    kernel_data_start = kernel_block.find("{")
    kernel_data_end = kernel_block.find("};") + 1
    kernel_raw = kernel_block[kernel_data_start:kernel_data_end]

    # Clean and parse nested array
    kernel_raw = kernel_raw.replace("\n", "").replace(" ", "")
    kernel_raw = kernel_raw.replace("{", "[").replace("}", "]")
    kernels = eval(kernel_raw)  # Safe only if file is trusted!
    kernels = [np.array(k) for k in kernels]

    # === Parse dense biases ===
    bias_start = content.find("static const float b_dense")
    if bias_start == -1:
        raise ValueError("b_dense block not found")

    bias_block = content[bias_start:]
    bias_data_start = bias_block.find("{")
    bias_data_end = bias_block.find("};")
    bias_raw = bias_block[bias_data_start + 1 : bias_data_end]
    b_dense = np.array([[float(x)] for x in bias_raw.split(",") if x.strip()])

    # === Parse dense weights ===
    W_start = content.find("static const float W_dense")
    if W_start == -1:
        raise ValueError("W_dense block not found")

    W_block = content[W_start:]
    W_data_start = W_block.find("{")
    W_data_end = W_block.find("};")
    W_lines = W_block[W_data_start + 1 : W_data_end].split("},")
    dense_weights = []
    for line in W_lines:
        line = line.strip().lstrip("{").rstrip("}")
        nums = [float(x.strip()) for x in line.split(",") if x.strip()]
        if nums:
            dense_weights.append(nums)

    W_dense = np.array(dense_weights)

    return kernels, W_dense, b_dense


# === Activation functions ===

def relu(x): return np.maximum(0, x)

def softmax(x):
    e_x = np.exp(x - np.max(x, axis=0))
    return e_x / np.sum(e_x, axis=0)

# === Inference ===
def predict(image, kernels, W_dense, b_dense):
    feature_maps = [relu(convolve2d(image, k, mode='valid')) for k in kernels]
    stacked = np.array(feature_maps)  # shape: (num_kernels, 26, 26)
    flat = stacked.reshape(-1, 1)
    Z = W_dense @ flat + b_dense
    A = softmax(Z)
    return np.argmax(A), A


def parse_header_file(path):
    with open(path, 'r') as f:
        lines = f.readlines()

    # Extract the label
    label_line = [line for line in lines if 'static const int test_label' in line][0]
    label = int(re.search(r'\d+', label_line).group())

    # Extract the image array
    array_lines = []
    in_array = False
    for line in lines:
        if 'static const float test_image' in line:
            in_array = True
            continue
        if in_array:
            if '};' in line:
                break
            array_lines.append(line.strip().strip(','))
    
    # Parse rows
    image = []
    for row in array_lines:
        numbers = [float(num) for num in row.strip('{}').split(',')]
        image.append(numbers)
    
    return np.array(image, dtype=np.float32), label


# === Test ===

if __name__ == "__main__":
    from matplotlib import pyplot as plt
    import pandas as pd

    # Load model
    kernels, W_dense, b_dense = load_cnn_model("cnn_model.h")

    # Load image and true label from C header
    image, label = parse_header_file("test_image_data.h")

    # Reshape image if needed
    image = image.reshape((28, 28))

    # Predict using the model
    predicted_label, probs = predict(image, kernels, W_dense, b_dense)

    # Print predicted, true label, and probabilities
    print(f"Predicted: {predicted_label}, True label: {label}")
    print("Probabilities per digit:")
    for digit, prob in enumerate(probs.ravel()):
        print(f"Digit {digit}: {prob:.6f}")

    # Show the image
    import matplotlib.pyplot as plt
    plt.imshow(image, cmap="gray")
    plt.title(f"Label from header: {label}")
    plt.show()
